### PR TITLE
[GitUp] libgit v1.1.0

### PR DESCRIPTION
**Pull request by lolgear**
# NB: WIP

# Overview

If you check commits in branch `gitup`, you may notice, that some of them are merged into current libgit main branch. However, these commits have different commits hashes, so, it is easy to search via commit message.

Everything is merged prior to 694eff30a0 Added Xcode script.

These custom commits could be divided into three groups.
1. Already merged into libgit2 master.
2. Patch
3. Already merged into libgit2 but methods signatures are slightly different and patch still required with invocation of original method from libgit2 inside.

These custom commits could be divided by functionality.

1. libssh2 fixes.
2. libgit2 enhancements.

## Suggestion
I suggest to extract these methods into gitup extensions files. They have a prefix `gitup_`.
For `submodule.{h,c}` these files will be `gitup_submodule.{h,c}`.

Also these extensions will be imported by `gitup_extensions.h` header.

## Commits and groups

Done | Required | Group | Commit
------| ---- | ------ | -------
[/] | Unknown | libssh2 | f29d369a6 Allow passing invalid SSH key files
[/] | Required | libssh2 | 6f7b5f453 Configure a 30 seconds timeout for recv() and send() on new sockets
[/] | Required | libssh2 | 160885ff2 Fixed initialization of libssh2 sessions and set a default 30s timeout
[/] | Unknown | libssh2 | f826478dd Ensure errors returned by SSH and HTTP authentication callbacks are preserved
[x] | Unknown | libssh2 | 6722c185c Revert "Find libssh2 via pkg-config"
[ ] | Unknown | macOS missing API | 28c24ce63 Revert "p_futimes: support using futimens when available"
[/] | Required | Patch (Possible in main) | fc0392bbc Pass current remote to git_push_negotiation callback
[/] | Unknown | Patch (Possible in main) | 345fcee9e Added git_repository_update_gitlink() API
[/] | Unknown | Patch (Possible in main) | 4d4d2ead5 Added git_clone_into() API
[/] | Upstream PR | Patch (Possible for PR) | 9c912cfd7 Added git_branch_upstream_merge()
[x] | Upstream PR | Patch (Hard) | Exposed new SPI git_branch_upstream_name_from_merge_remote_names()
[/] | Unknown | Patch (Unknown) | 487c1a689 Added git_config_find_local()
[x] | Unknown | Patch (Unknown) | a83a27492 Add time metadata to git_diff_file
[/] | Upstream PR | Patch (Possible in main) | 8b6f0abe1 Added virtual references API
[x] | Upstream PR | Patch (Simple) | 57d4e3052 Added git_submodule_retain()
[x] | Required | Xcode | 694eff30a Added Xcode script

### CODE REVIEW

**Comment by lucasderraugh** [comment]
This does not contain all the changes Pierre made in this fork. A simple rebase onto this branch yields many more conflicts and this doesn't capture all of what is missing. Ultimately to open a PR against this there needs to be a corresponding GitUp branch that integrates these changes and passes all the tests. I'm pretty sure we're missing tons of functions, one such example would be `git_branch_upstream_merge`.

**Comment by lolgear** [comment]
This PR is still in progress.
I write down all my thoughts about current state of this patch ( how to structure it and how to handle all commits ).
As an example, it contains several commits with new structure of patches ( all functions are extracted to their headers and all files and all functions are prefixed with gitup_ ).

**Comment by lolgear** [comment]
It may be ready, but I think it requires a branch in GitUp with renaming functions. ( `git_` -> `gitup_` ).
It is possible that some changes will be merged in libgit2, so, it will be easy to rebase next time.

**Comment by lucasderraugh** [comment]
@lolgear The first thing would be to open up a new GitUp branch then and update the submodule to point to this commit and test it.

**Comment by lolgear** [comment]
I opened PR with updated libgit2 in GitUp repository.
It is too rough now, because I have encountered obstacles.

I would like to discuss an update of libgit2 project in GitUp third-party directory.

It contains a libgit2 target and dependencies ( libssh2 ).
Also it contains xconfigs.

I think that it would be better if these things are handled semi-automatic by scripts.
But I don’t remember if adding dynamic library could be automatic.

At this circumstances, I think about additional script that will call cmake and add these dynamic libraries and set build configurations ( if possible ).

All steps could be

1. Call `update_xcode.sh` to build libgit2.xcproject.
2. Provide libssh2 support feature by setting macros variable in features.h.
3. Add dynamic libraries to targets.
4. Add xcconfigs.

All these steps may be written in cmake. ( I wish it contains these functions ).

Also I don’t think whether `xcconfigs` are required in this solution.

What do you think about it?

—
Best regards,
Dmitry

**Comment by lolgear** [comment]
I'm thinking about another "modern" approach.
Maybe it is more convenient to add Package.swift to this fork.
Yes, this `Package.swift` will take time to configure properly.
But it eliminates `libgit2.xcodeproj` and xcode script is not required anymore. ( Well, it is required for generating "features" header with definitions of feature-flags ).

One thing that can block this approach is `xcframework` dependency.
All these libgit2 targets have local xcframework dependencies ( libssh2, libopenssl ).

This package (Package.swift) will be a part of GitUp repository, so, we don't need to add Package.swift in this fork.

@lucasderraugh 
What do you think?

**Comment by lucasderraugh** [comment]
I don't see how a Package.swift is going to help us. We aren't using Swift in the main project and won't be for the foreseeable future. We should use the same approach as the Objective-git project: https://github.com/libgit2/objective-git.

Whatever we'd do I don't want to introduce Swift Package Manager or CocoaPods. We should stick with the submodule approach we're using today. @lolgear

**Comment by lolgear** [comment]
@lucasderraugh 
CMake has several problematic parts.
1. It doesn't use relative paths for source files.
2. It doesn't use Xcode settings at all and move everything into Other C Flags/Other linker flags.
3. It can't add xcframework dependencies.

Current Xcodeproj file is outdated and it is very hard to adjust settings from autogenerated Xcodeproj file to a current Xcodeproj file.

SwiftPM solves all problems that I mention above. 
Also it produces static C-libraries.

SwiftPM here is just a new format for targets and nothing more.
Submodules will exist. 
GitUp will hook to Package.swift file and use it to get configured C-library.
Basically, SwiftPM replaces libgit2.xcodeproj.

If you update main branch, I will add new PR and show you have it works together.

**Comment by lucasderraugh** [comment]
@lolgear By "update main branch" you just mean merge this PR into `gitup_libgit_v1.1` branch?

**Comment by lolgear** [comment]
@lucasderraugh 
Pull changes from upstream.
I use this commit for new branch.
https://github.com/libgit2/libgit2/commit/e65229ee972c113413eeca77853213352129bd47

**Comment by lucasderraugh** [comment]
@lolgear I've updated it to libgit2's main branch. I want to be clear though that we cannot point to a non-release version of libgit2 for release. So ultimately before anything would go in to a release version of GitUp I'd have to wait until 1.1.2 (or whatever the next version is).

That said, you can use this to demonstrate your changes against with that commit you listed above, but we'll have to wait before GitUp will get this change.
